### PR TITLE
Fix unreleased form locks

### PR DESCRIPTION
**Pull request by millerdev**
Duplicate processing has been moved into `FormProcessingResult.get_locked_forms()` to avoid acquiring any form locks before ~that method is called~ the returned context manager is entered. There is now only one lock on the (possibly duplicate) submitted form ID. All other form IDs involved in the processing are newly generated UUIDs and therefore should not have contention unless there is a collision (and that should be suitably rare to not be a concern).

This should resolve the following:

Consider [this code block](https://github.com/dimagi/commcare-hq/blob/0c65b8424f39f3a9ea8c7e324db06acb05674987/corehq/form_processor/submission_post.py#L219-L237). The lock is acquired on line 219 (previously I had thought it was acquired on line 237). The two return statements after line 219 but before line 237 happen before the lock is released. In other words, the lock is acquired but never released (until timeout).

@snopoke @dannyroberts @ctsims

### CODE REVIEW

File: corehq/form_processor/parsers/form.py [hunk 1]
```diff
@@ -10,75 +12,94 @@
...
+                    if xform_id == new_id:
+                        manager.acquire_lock(old_id)
+                    elif xform_id == old_id:
+                        manager.acquire_lock(new_id)
+                    else:
+                        # unexpected, should never get here
+                        assert False, (xform_id, new_id, old_id)
+                elif new_id != xform_id:
+                    manager.context = [new_form]
+                    manager.acquire_lock(new_id)
```
Comment (snopoke) [inline corehq/form_processor/parsers/form.py]: Do we even need a lock here? In this case the `new_id` has just been assigned to the form on the server so unless there's a UUID clash we shouldn't need to worry about another form being submitted with the same ID.

Comment (snopoke) [reply corehq/form_processor/parsers/form.py]: The same applies to the code above.

Comment (millerdev) [reply corehq/form_processor/parsers/form.py]: Good point. I just tried to stick to what the old code was doing. We may be able to remove these locks in the future, but I don't think they will hurt anything if they're released reliably.

Comment (millerdev) [reply corehq/form_processor/parsers/form.py]: e8c86fc

**Review by dannyroberts — APPROVED** [review]
Took me a while to wrap my head around it, but the added description was helpful, and this makes a lot of sense. Thanks for digging into it and finding this nice fix

File: corehq/form_processor/parsers/form.py [hunk 2]
```diff
@@ -10,75 +11,55 @@
...
+@contextmanager
+def locked_form(xform, interface):
+    context = [xform]
+    lock = interface.acquire_lock_for_xform(xform.form_id)
+    try:
+        if interface.is_duplicate(xform.form_id):
+            new_form, dup_form = _handle_id_conflict(xform, xform.domain)
+            if dup_form:
+                assert dup_form.form_id != new_form.form_id, (new_form, dup_form)
+                context[:] = [new_form, dup_form]
```
[the remark this answers is not in the document]
Comment (snopoke) [reply corehq/form_processor/parsers/form.py]: I see you changed it in https://github.com/dimagi/commcare-hq/pull/22324/files

**Pull request merged**